driver-core/src/main/com/mongodb/internal/async/function/RetryState.java

```diff
@@ -203,15 +193,6 @@ private void doAdvanceOrThrow(final Throwable attemptException,
          */
         if (isLastAttempt() || attemptException instanceof MongoOperationTimeoutException) {
             previouslyChosenException = newlyChosenException;
-            /*
-             * The function of isLastIteration() is to indicate if retrying has been explicitly halted. Such a stop is not interpreted as
-             * a timeout exception but as a deliberate cessation of retry attempts.
-             */
-            if (hasTimeoutMs() && !loopState.isLastIteration()) {
```

Comment:
The logic for removing these is that it is safe to behave as if a timeout has not expired (even if it has), because it will be detected and thrown by an ensuing blocking operation.

This relies on the assumption that all blocking operations will throw when given an expired timeout (this should be the case, because all blocking operations should now be wrapped in Timeout's branching "call" methods).

Reply:
Tests were failing. The logic related to timeout exceptions in RetryState is necessary, but complicated to extract. I've restored it, but the isExpired check is still removed (so this still relies on the ensuing blocking operation throwing a timeout exception).